### PR TITLE
[PEFT] Allow PEFT model dict to be loaded

### PR DESCRIPTION
**Pull request by patrickvonplaten**
In order to allow `peft` to be leveraged in `diffusers` without breaking changes we need to allow loading adapters directly from a loaded `state_dict`. The reason is that in `diffusers` we currently store LoRA checkpoints in a format that is different to the PEFT format so we cannot just pass the model_id. This PR allows the user to manually pass a loaded PEFT model checkpoint as well as a PEFT configuration, thus circumventing the need to pass a model id.

In pseudo code, the integration of `transformers` + PEFT in `diffusers` should then look as follows the ["load_lora"](https://github.com/huggingface/diffusers/blob/24c5e7708bb75076dd8e79ccaea195640555f945/src/diffusers/loaders.py#L1213) function of `diffusers`.

```py
def load_lora_into_text_encoder(cls, state_dict, network_alphas, text_encoder, prefix=None, lora_scale=1.0):
    peft_state_dict, peft_config = convert_to_peft_format(state_dict, ...)  # <- this function will take care of all the remapping necessary for the different formats
    text_encoder.load_adapter(peft_state_dict, peft_config=peft_config)
```

**Note**, there might be more changes we have to do to PEFT, `transformers'` PEFT integration to be sure that everything works as expected. E.g. I'm not yet sure how to pass `network_alphas` etc... to PEFT to make sure we get 1-to-1 the same result.

cc @younesbelkada @sayakpaul @BenjaminBossan @pacman100

### CODE REVIEW

File: src/transformers/lib_integrations/peft/peft_mixin.py [hunk 1]
```diff
@@ -59,14 +64,15 @@ class PeftAdapterMixin:
 
     def load_adapter(
         self,
-        peft_model_id: str,
+        peft_model_id: Union[str, Dict[str, torch.Tensor]],
```
Comment (patrickvonplaten) [inline src/transformers/lib_integrations/peft/peft_mixin.py:67]:
```suggestion
        peft_model_id: Union[str, Dict[str, "torch.Tensor"]],
```

File: src/transformers/lib_integrations/peft/peft_mixin.py [hunk 2]
```diff
@@ -28,6 +29,10 @@
     from accelerate.utils import get_balanced_memory, infer_auto_device_map
 
 
+if is_torch_available():
```
Comment (BenjaminBossan) [inline src/transformers/lib_integrations/peft/peft_mixin.py:32]: Why is this necessary?

File: src/transformers/lib_integrations/peft/peft_mixin.py [hunk 3]
```diff
@@ -59,14 +64,15 @@ class PeftAdapterMixin:
 
     def load_adapter(
         self,
-        peft_model_id: str,
+        peft_model_id: Union[str, Dict[str, "torch.Tensor"]],
```
Comment (BenjaminBossan) [inline src/transformers/lib_integrations/peft/peft_mixin.py:67]: Should we choose a different name, now that a state dict can be passed? Alternatively, we could add another (optional) argument to pass the state dict, make `peft_model_id` optional, and do a check that exactly one of the two should be passed.

File: src/transformers/lib_integrations/peft/peft_mixin.py [hunk 4]
```diff
@@ -75,7 +81,7 @@ def load_adapter(
         Requires peft as a backend to load the adapter weights.
 
         Args:
-            peft_model_id (`str`):
+            peft_model_id (`str` or dictionary of `torch.Tensor`):
```
Comment (BenjaminBossan) [inline src/transformers/lib_integrations/peft/peft_mixin.py:84]: The description below should also be adjusted.

File: src/transformers/lib_integrations/peft/peft_mixin.py [hunk 5]
```diff
@@ -59,14 +64,15 @@ class PeftAdapterMixin:
...
-        peft_model_id: str,
+        peft_model_id: Union[str, Dict[str, "torch.Tensor"]],
         adapter_name: Optional[str] = None,
         revision: Optional[str] = None,
         token: Optional[str] = None,
         device_map: Optional[str] = "auto",
         max_memory: Optional[str] = None,
         offload_folder: Optional[str] = None,
         offload_index: Optional[int] = None,
+        peft_config: Dict[str, Any] = None,
```
Comment (BenjaminBossan) [inline src/transformers/lib_integrations/peft/peft_mixin.py:75]:
```suggestion
        peft_config: Optional[Dict[str, Any]] = None,
```

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for the PR.

I'm not very knowledgeable about diffusers, so will let others comment on the overall solution. My comments are just minor issues, not blockers.

File: src/transformers/lib_integrations/peft/peft_mixin.py [hunk 6]
```diff
@@ -125,28 +131,32 @@ def load_adapter(
...
-            peft_model_id,
-            revision=revision,
-            token=token,
-        )
-
-        if adapter_config_file is None:
-            raise ValueError(
-                f"adapter model file not found in {peft_model_id}. Make sure you are passing the correct path to the "
-                "adapter model."
+        if peft_config is None:
```
Comment (younesbelkada) [inline src/transformers/lib_integrations/peft/peft_mixin.py:134]:
```suggestion
        if peft_config is None and isinstance(peft_model_id, str):
```
and add a check below that if peft_model_id is not a state dict, raise an error

**Review by younesbelkada — APPROVED** [review]
Looks great! agreed also with @BenjaminBossan 's comments, I think we should maybe add an extra check and raise a proper error, for the naming, what about `peft_model_id_or_state_dict` (maybe that's too long) ?

File: src/transformers/lib_integrations/peft/peft_mixin.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/lib_integrations/peft/peft_mixin.py:67]: Yes makes sense to me! `peft_model_id_or_state_dict` sounds great - was just wondering about breaking change here. But think this function is not yet in a release so should probs be fine to change no?

**Review by pacman100 — COMMENTED** [review]
Thank you @patrickvonplaten for adding the support for passing state dict and config, LGTM!

**Review by pacman100 — APPROVED** [review]

**Review by younesbelkada — APPROVED** [review]
Thanks again !

File: src/transformers/integrations/peft.py [hunk 7]
```diff
@@ -77,7 +83,7 @@ def load_adapter(
         Requires peft as a backend to load the adapter weights.
 
         Args:
-            peft_model_id (`str`):
+            peft_model_id (`str` or dictionary of `torch.Tensor`):
                 The identifier of the model to look for on the Hub, or a local path to the saved adapter config file
                 and adapter weights.
```
Comment (ArthurZucker) [inline src/transformers/integrations/peft.py:88]: Doc needs to be updated if this can be a ` state_dict`. The name of the argument is not intuitive but I guess wee need to keep it for Backward compatibility? Otherwise would rather have a new arg

File: src/transformers/integrations/peft.py [hunk 8]
```diff
@@ -61,14 +66,15 @@ class PeftAdapterMixin:
...
-        peft_model_id: str,
+        peft_model_id: Union[str, Dict[str, "torch.Tensor"]],
         adapter_name: Optional[str] = None,
         revision: Optional[str] = None,
         token: Optional[str] = None,
         device_map: Optional[str] = "auto",
         max_memory: Optional[str] = None,
         offload_folder: Optional[str] = None,
         offload_index: Optional[int] = None,
+        peft_config: Dict[str, Any] = None,
```
Comment (ArthurZucker) [inline src/transformers/integrations/peft.py:77]: not in docstring

File: src/transformers/integrations/peft.py [hunk 9]
```diff
@@ -127,28 +133,32 @@ def load_adapter(
...
 
         # Create and add fresh new adapters into the model.
-        inject_adapter_in_model(loaded_peft_config, self, adapter_name)
+        inject_adapter_in_model(peft_config, self, adapter_name)
 
-        adapter_state_dict = load_peft_weights(peft_model_id, revision=revision, use_auth_token=token)
+        if not isinstance(peft_model_id, dict):
+            adapter_state_dict = load_peft_weights(peft_model_id, revision=revision, use_auth_token=token)
+        else:
+            adapter_state_dict = peft_model_id
```
Comment (ArthurZucker) [inline src/transformers/integrations/peft.py:161]: Just checking that we can' t add a new ` adapter_state_dict`  arg right?

File: src/transformers/integrations/peft.py [hunk 10]
```diff
@@ -30,6 +31,10 @@
 # Minimum PEFT version supported for the integration
 MIN_PEFT_VERSION = "0.5.0"
 
+if is_torch_available():
+    import torch
+
```
Comment (ArthurZucker) [inline src/transformers/integrations/peft.py:36]:
```suggestion
if TYPE_CHECKING:
    if is_torch_available():
        import torch  # noqa
```
as it it' s not directly used

**Review by ArthurZucker — COMMENTED** [review]
Have one question regarding the ` peft_model_id`  being a dictionary, otherwise looks good

File: src/transformers/integrations/peft.py (hunk 7 above)
Comment (younesbelkada) [reply src/transformers/integrations/peft.py:88]: I think for BC it's fine as the feature is quite new

File: src/transformers/integrations/peft.py (hunk 10 above)
Comment (younesbelkada) [reply src/transformers/integrations/peft.py]: done in [55d657b](https://github.com/huggingface/transformers/pull/25721/commits/55d657b072beddf4080cb2b44c0b33fafb8f7e87)

File: src/transformers/integrations/peft.py (hunk 8 above)
Comment (younesbelkada) [reply src/transformers/integrations/peft.py:78]: [55d657b](https://github.com/huggingface/transformers/pull/25721/commits/55d657b072beddf4080cb2b44c0b33fafb8f7e87)

File: src/transformers/integrations/peft.py (hunk 9 above)
Comment (ArthurZucker) [reply src/transformers/integrations/peft.py]: Why?

File: src/transformers/integrations/peft.py (hunk 7 above)
Comment (ArthurZucker) [reply src/transformers/integrations/peft.py]: Not really if it was part of the release 😅

File: src/transformers/integrations/peft.py (hunk 9 above)
Comment (younesbelkada) [reply src/transformers/integrations/peft.py]: Actually it turns out it is possible, please check [d05ef41](https://github.com/huggingface/transformers/pull/25721/commits/d05ef41d86c8f9566ac4977e752dd847e222daef)

File: tests/peft_integration/test_peft_integration.py [hunk 11]
```diff
@@ -300,3 +302,32 @@ def test_peft_pipeline(self):
...
+
+                with self.assertRaises(ValueError):
+                    model.load_adapter(peft_model_id=None)
+
+                state_dict_path = hf_hub_download(peft_model_id, "adapter_model.bin")
+
+                dummy_state_dict = torch.load(state_dict_path)
+
+                model.load_adapter(adapter_state_dict=dummy_state_dict, peft_config=peft_config)
+
```
Comment (ArthurZucker) [inline tests/peft_integration/test_peft_integration.py:329]:
```suggestion
                model.load_adapter(adapter_state_dict=dummy_state_dict, peft_config=peft_config)
                with self.assertRaises(ValueError):
                    model.load_adapter(model.load_adapter(adapter_state_dict=dummy_state_dict, peft_config=None))
```
this should raise an error as well !

**Review by ArthurZucker — APPROVED** [review]
LGTM thanks for iterating and adding a test!
